rfd/0125-dynamic-auto-discovery-config.md

```diff
@@ -0,0 +1,406 @@
+Users no longer need to log in to a machine that is running this service to change the matchers, and can do any sort of configuration from the clients (ie WebApp, `tctl`, Terraform Provider and Kubernetes Operator).
+
+## Details
+There will be a new resource - `DiscoveryConfig` - which will be used by the `discovery_service` to create a set of matchers for that particular service.
+
+We'll create relations between those two entities using labels and label matchers.
+This is common for other type of services, eg `db_service` has the `resources.labels` labels matcher which allows users to define which Databases to proxy based on their labels.
+
+### New `DiscoveryConfig` resource
+A new resource will be created to store the desired matchers.
+
+The `spec` schema of the `DiscoveryConfig` contains all the required fields for matching on resources on supported clouds.
+
+Currently, the schema is compose using the following matchers:
+- `lib/config.AWSMatcher`
+- `lib/config.AzureMatcher`
+- `lib/config.GCPMatcher`
+
+In order to use those types, we must first define them as protobuf types and then import them in `lib/config`.
+
+This way, we have a single place where we define the matchers properties.
+They are then used for file reading or for defining `DiscoveryConfig` resources.
+
+Example:
+```yaml
+kind: DiscoveryConfig
+version: v1
+metadata:
+  name: production-resources
+  labels:
+    cloud: aws
+    env: prod
+spec:
+    aws:
+     - types: ["ec2"]
+       regions: ["us-east-1","us-west-1"]
+       tags:
+         "*": "*"
+       install:
+         join_params:
+           token_name:  "aws-discovery-iam-token"
+         script_name: "default-installer"
+       ssm:
+         document_name: "TeleportDiscoveryInstaller"
+    azure:
+    - types: ["aks"]
+      regions: ["eastus", "westus"]
+      subscriptions: ["11111111-2222-3333-4444-555555555555"]
+      resource_groups: ["group1", "group2"]
+      tags:
+        "*": "*"
+    gcp:
+    - types: ["gke"]
+      locations: ["*"]
+      tags:
+        "*": "*"
+      project_ids: ["myproject"]
+```
+
+### Discovery Group
+The Discovery Group property will not be ported to `DiscoveryConfig` because it exists to ensure that multiple `discovery_services` don't step on each other.
+
+Running `discovery_service`s in high availability (multiple instances and same config) should yeld the same result as before.
+
+### New `discovery_service` property: `selector`
+The configuration of the `discovery_service` has a new field: `selector`.
+
+This field works as a label matcher for `DiscoveryConfig` resources.
+Standard label matcher rules apply here (eg, Role's `node_labels` or Database Service's `resources.labels`).
+
+For each matching `DiscoveryConfig`, the list of matchers is appended to the `discovery_service` matchers.
+
+Example:
+```yaml
+discovery_service:
+  enabled: "yes"
+  discovery_group: my_group
+  selector:
+    cloud: aws
+    env: [qa, prod]
```

Comment:
This should be a list.

Also, I don't think we support `env: [qa, prod]` syntax. Users would need to specify multiple selectors (which is why we need a list).

Reply:
`selector` will have `types.Labels` type
https://github.com/gravitational/teleport/blob/e55f844f3b1af2f6283e142eb7c787fdd75e4a98/api/types/role.go#L1107-L1110

Which supports either a single element or a list.

So, something like this (which was my inspiration)
https://github.com/gravitational/teleport/blob/05aa0488a178292cfa384b491589d54934d4c76b/lib/services/databaseservice_test.go#L71-L80